src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+        resp = COMMANDS[0][2]()
+        resp.size = 65
+        resp.signing = NEGOTIATE_SIGNING_ENABLED
+        resp.dialect = self.dialect
+        resp.server_uuid = self.factory.server_uuid.bytes_le
+        resp.capabilities = GLOBAL_CAP_DFS
+        resp.max_transact = MAX_TRANSACT_SIZE
+        resp.max_read = MAX_READ_SIZE
+        resp.max_write = MAX_WRITE_SIZE
+        resp.time = base.u2nt_time(time.time())
+        resp.boot_time = base.u2nt_time(self.factory.server_start)
+        resp.offset = 128
+        resp.buflen = len(blob)
+        resp.buffer = blob
+        self.send_with_header(resp, 'negotiate')
+
+    def smb_session_setup(self, req, resp):
+        blob = req.buffer[req.offset - len(req) - 64:req.offset - len(req) -
+                          64 + req.buflen]
+        log.debug("SESSION SETUP")
+        log.debug("-------------")
+        log.debug("Size             {sz}", sz=req.size)
+        log.debug("Security mode    0x{sm:08x}", sm=req.security_mode)
+        log.debug("Capabilities     0x{cap:08x}", cap=req.capabilities)
+        log.debug("Channel          0x{chl:08x}", chl=req.channel)
+        log.debug("Prev. session ID 0x{pid:016x}", pid=req.prev_session_id)
+        if self.first_session_setup:
+            self.blob_manager.receiveInitialBlob(blob)
+            resp.buffer = self.blob_manager.generateChallengeBlob()
+            self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+            self.first_session_setup = False
+        else:
+            self.blob_manager.receiveResp(blob)
+            if self.blob_manager.credential:
+                log.debug("got credential: %r" % self.blob_manager.credential)
+                d = self.factory.portal.login(
+                    self.blob_manager.credential,
+                    SMBMind(req.prev_session_id,
+                            self.blob_manager.credential.domain, self.addr),
+                    ISMBServer)
+                d.addCallback(self._cb_login, resp)
+                d.addErrback(self._eb_login, resp)
+            else:
+                resp.buffer = self.blob_manager.generateChallengeBlob()
+                self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+
+    def _cb_login(self, t, resp):
+        _, self.avatar, self.logout_thunk = t
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(True)
+        log.debug("successful login")
+        self.session_setup_response(resp, STATUS_SUCCESS)
+
+    def _eb_login(self, failure, resp):
+        log.debug(failure.getTraceback())
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(False)
+        self.session_setup_response(resp, STATUS_LOGON_FAILURE)
+
+    def session_setup_response(self, resp, ntstatus):
+        log.debug("session_setup_response")
+        resp.flags = 0
+        if self.blob_manager.credential == ANONYMOUS:
+            resp.flags |= SESSION_FLAG_IS_NULL
+        resp.size = 9
+        resp.offset = 72
```

Comment:
`72` is one of the more magic values I've seen so far here :)

Reply:
Each packet type has two "magic" numbers, a request size and a response size, some have a third, the "offset". For reasons I can't fathom they're true magic: nearly but not quite the same as the size of the packet on the wire. If we move to `attrs` based pack/unpack these can be set as defaults in the class definition.

All other instances of "magic" values should be factored out to named constants now.